src/storage/v2/compact_vector.hpp

```diff
@@ -315,13 +190,237 @@ struct CompactVector {
+        size_ = other.size_;
+      } else if (other.size_ < size_) {
+        // destroy
+        std::destroy(dst_assign_end, end());
+        size_ = other.size_;
+      }
+    }
+    return *this;
+  }
+
+  auto operator=(CompactVector &&other) noexcept -> CompactVector & {
+    if (std::addressof(other) != this) {
+      if (usingSmallBuffer(other.capacity_)) {
+        assert(other.capacity_ <= capacity_);
+        // move assignment
+        auto to_assign = std::min(size_, other.size_);
+        auto src_assign_end = other.begin() + to_assign;
+        auto dst_assign_end = begin() + to_assign;
+        std::move(other.begin(), src_assign_end, begin() /*, dst_assign_end*/);
+
+        if (size_ < other.size_) {
+          // move construct
+          auto dst_ctr_end = begin() + other.size_;
+          std::ranges::uninitialized_move(src_assign_end, other.end(), dst_assign_end, dst_ctr_end);
+          size_ = other.size_;
+        } else if (other.size_ < size_) {
+          // destroy
+          std::destroy(dst_assign_end, end());
+          size_ = other.size_;
+        }
+      } else {
+        if (usingSmallBuffer(capacity_)) {
+          std::destroy(begin(), end());
+          size_ = std::exchange(other.size_, 0);
+          capacity_ = std::exchange(other.capacity_, kSmallCapacity);
+          buffer_ = other.buffer_;
+        } else {
+          std::destroy(begin(), end());
+          size_ = std::exchange(other.size_, 0);
+          capacity_ = std::exchange(other.capacity_, kSmallCapacity);
+          auto old_buffer = std::exchange(buffer_, other.buffer_);
+          std::free(old_buffer);
+        }
+      }
+    }
+    return *this;
+  }
+
+  ~CompactVector() {
+    std::destroy(begin(), end());
+    if (!usingSmallBuffer(capacity_)) {
+      std::free(buffer_);
+    }
+  }
+
+  explicit CompactVector(std::initializer_list<T> other) : CompactVector(other.begin(), other.end()) {}
+
+  // TODO: change to range + add test
+  explicit CompactVector(std::span<T const> other) : CompactVector(other.cbegin(), other.cend()) {}
+  // TODO: generalise to not just vector
+  explicit CompactVector(std::vector<T> &&other) : size_(other.size()), capacity_{std::max(size_, kSmallCapacity)} {
+    if (!usingSmallBuffer(capacity_)) {
+      buffer_ = reinterpret_cast<pointer>(std::aligned_alloc(alignof(T), capacity_ * sizeof(T)));
+    }
+    std::ranges::uninitialized_move(other.begin(), other.end(), begin(), end());
+  }
+
+  template <typename It>
+  explicit CompactVector(It first, It last)
+      : size_(std::distance(first, last)), capacity_{std::max(size_, kSmallCapacity)} {
+    if (!usingSmallBuffer(capacity_)) {
+      buffer_ = reinterpret_cast<pointer>(std::aligned_alloc(alignof(T), capacity_ * sizeof(T)));
+    }
+    std::ranges::uninitialized_copy(first, last, begin(), end());
+  }
+
+  void push_back(const_reference value) {
+    if (capacity_ == size_) {
+      reserve(capacity_ == 0 ? 1 : 2 * capacity_);
```

Comment:
I think `capacity_` now can't be 0. It's minimum `kSmallCapacity`.

Reply:
It can be 0 if `8 < sizeof(T)`